File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        char start;
        if(needle.size()){
            start=needle[0];
        }
        else return 0;
        int idx=-1;
        for(int i=0;i<haystack.size();i++){
            if(haystack[i]==start){
                int count=0,j,k=0;
                for(j=i;j<haystack.size();j++){
                    if(haystack[j]==needle[k])count++;
                    else break;k++;
                }
                if(count==needle.size())return i;
                // i=j-1;
            }
        }
        return -1;
    }
};

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.java
class Solution {
    public int strStr(String haystack, String needle) {
         int n = haystack.length();
        int m = needle.length();
        if (m == 0) return 0;
        int[] pi = new int[m];
        pi[0] = 0;
        int j = 0;
        for (int i = 1; i < m; i++) {
            while (j > 0 && needle.charAt(j) != needle.charAt(i)) {
                j = pi[j - 1];
            }
            if (needle.charAt(j) == needle.charAt(i)) {
                j++;
            }
            pi[i] = j;
        }
        j = 0;
        for (int i = 0; i < n; i++) {
            while (j > 0 && needle.charAt(j) != haystack.charAt(i)) {
                j = pi[j - 1];
            }
            if (needle.charAt(j) == haystack.charAt(i)) {
                j++;
            }
            if (j == m) {
                return i - m + 1;
            }
        }
        return -1;
    }
}